File: gateway/rules.py
"""Deterministic guardrails for the Crown-Jewels Gateway.

These run BEFORE the model. Anything matched here is forced to PRIVATE so a
model misjudgment can never leak a crown jewel (belt-and-suspenders).
"""
import re

_PATTERNS = {
    # Currency symbol/code + number, OR number + currency/magnitude word, OR a
    # bare magnitude word — catches "Rs 45,000", "$2.5M", "5000 dollars", and
    # "lost millions in sales".
    "money": re.compile(
        r"(₹|rs\.?|inr|usd|eur|gbp|\$|€|£)\s?\d[\d,.]*"
        r"|\b\d[\d,.]*\s?(k|m|bn|dollars?|rupees?|euros?|pounds?|lakhs?|crores?)\b"
        r"|\b(millions?|billions?|trillions?|lakhs?|crores?)\b",
        re.I,
    ),
    # Business crown jewels — err PRIVATE (safe default).
    "finance": re.compile(
        r"\b(revenue|sales|profit|turnover|salary|payroll|invoice|earnings"
        r"|valuation|funding|acquisition|net\s?worth)\b",
        re.I,
    ),
    "card": re.compile(r"\b(?:\d[ -]?){13,16}\b"),
    "account": re.compile(r"\b(a/c|acct|account)\b.*\d{4,}", re.I),
    "phone": re.compile(r"\b(?:\+?91[- ]?)?[6-9]\d{9}\b"),
    "health": re.compile(r"\b(diagnos|prescription|medical|blood test|report)\b", re.I),
}

# Populated from the UI privacy toggles.
DEFAULT_PRIVATE_KEYWORDS = ["salary", "loan", "emi", "family", "school", "client"]


def rule_hits(text, private_keywords=None):
    """Return the list of rule categories that force this item PRIVATE.

    private_keywords semantics:
      None  -> use DEFAULT_PRIVATE_KEYWORDS (no UI policy supplied).
      []    -> no keyword overrides (every toggle is off); patterns still apply.
      [...] -> exactly this policy, from the UI privacy toggles.
    """
    if private_keywords is None:
        private_keywords = DEFAULT_PRIVATE_KEYWORDS
    # Coerce to str: this is the can't-miss guardrail, so a non-string item
    # (int, list, None from a malformed ingest) must degrade to "no text",
    # never crash the regex engine on `.search()` / `.lower()`.
    text = "" if text is None else str(text)
    hits = [name for name, pat in _PATTERNS.items() if pat.search(text)]
    # Keyword hits are case-insensitive WORD-boundary matches, not substrings, so
    # "emi" fires on the loan term but not inside "reminder"/"premium". Still
    # catches the whole keyword anywhere in the text; mixed-case toggles work too.

    # Fast-path check: avoid regex overhead if keyword doesn't appear as a substring
    text_lower = text.lower()
    for kw in private_keywords:
        k = str(kw).lower()
        if k in text_lower and re.search(r"\b" + re.escape(k) + r"\b", text, re.I):
            hits.append(f"kw:{k}")
    return hits

File: extension/rules.js
/**
 * Deterministic privacy classifier — JS port of gateway/rules.py.
 *
 * Runs as pass 1 in the offscreen classifier (before Gemma 270M).
 * Any hit here forces PRIVATE unconditionally — a model misjudgment
 * can never leak a crown jewel (belt-and-suspenders).
 */

const PATTERNS = {
  // Currency symbol/code + number, OR number + currency/magnitude word,
  // OR a bare magnitude word (million/billion/lakh/crore) — catches
  // "Rs 45,000", "$2.5M", "5000 dollars", and "lost millions in sales".
  money:
    /(?:₹|rs\.?|inr|usd|eur|gbp|\$|€|£)\s?\d[\d,.]*|\b\d[\d,.]*\s?(?:k|m|bn|dollars?|rupees?|euros?|pounds?|lakhs?|crores?)\b|\b(?:millions?|billions?|trillions?|lakhs?|crores?)\b/i,
  // Business crown jewels — revenue, deals, comp. Err PRIVATE (safe default).
  finance:
    /\b(?:revenue|sales|profit|turnover|salary|payroll|invoice|earnings|valuation|funding|acquisition|net\s?worth)\b/i,
  card:    /\b(?:\d[ -]?){13,16}\b/,
  account: /\b(?:a\/c|acct|account)\b.*\d{4,}/i,
  phone:   /\b(?:\+?91[- ]?)?[6-9]\d{9}\b/,
  health:  /\b(?:diagnos|prescription|medical|blood\s?test|report)\b/i,
};

export const DEFAULT_PRIVATE_KEYWORDS = [
  'salary', 'loan', 'emi', 'family', 'school', 'client',
];

/**
 * Return the category names that force this text PRIVATE.
 * @param {string} text
 * @param {string[]} privateKeywords - user-configured extra keywords
 * @returns {string[]}
 */
export function ruleHits(text, privateKeywords = DEFAULT_PRIVATE_KEYWORDS) {
  const t = text == null ? '' : String(text);
  const hits = [];
  for (const [name, pat] of Object.entries(PATTERNS)) {
    if (pat.test(t)) hits.push(name);
  }
  // Word-boundary match, not substring: "emi" must fire on the loan term but
  // NOT inside "r-emi-nder" or "pr-emi-um". Still catches the keyword anywhere.
  for (const kw of privateKeywords) {
    const k = String(kw).toLowerCase();
    const re = new RegExp('\\b' + k.replace(/[.*+?^${}()|[\]\\]/g, '\\$&') + '\\b', 'i');
    if (re.test(t)) hits.push(`kw:${kw}`);
  }
  return hits;
}

/**
 * Full deterministic classification — used when WebGPU / Gemma is unavailable.
 * @param {string} text
 * @param {string[]} privateKeywords
 * @returns {{ tier: 'PRIVATE'|'SHARED', sensitivity: number, categories: string[], reason: string }}
 */
export function classifyFallback(text, privateKeywords = DEFAULT_PRIVATE_KEYWORDS) {
  const hits = ruleHits(text, privateKeywords);
  if (hits.length > 0) {
    return {
      tier: 'PRIVATE',
      sensitivity: 1.0,
      categories: hits,
      reason: 'matched deterministic rule(s)',
    };
  }
  return {
    tier: 'SHARED',
    sensitivity: 0.1,
    categories: [],
    reason: 'no rule hit (deterministic fallback)',
  };
}
